### ISSUE TITLE
UFP (Uplink Failure Detection) incomplete support 

### ISSUE THREAD

**Issue by prasadsriramps** [oopt-goldstone/goldstone-mgmt#77]
This issue discusses the UFD feature and how it is not disabling downlink ports based on the uplink port's operational state at the ASIC level. Once the user configures UFD uplink and downlink ports, south-sonic checks the operational state of uplink ports, and re-configures the admin state/sets the admin state to down in the REDIS Config DB, for the corresponding downlink port. 

During run-time, if the uplink port is operationally disabled, the event handler sets the operational status for downlink ports to "dormant", but the port still remains operational 'UP' at the hardware level. 
```
root@localhost:~# gscli
>
>
> show interface brief
+--------------+-------------+--------------+--------------+
|     name     | oper-status | admin-status |    alias     |
+--------------+-------------+--------------+--------------+
| Ethernet1_1  |     up      |      up      |  QSFP28E1-1  |
| Ethernet2_1  |     up      |      up      |  QSFP28E2-5  |
| Ethernet3_1  |    down     |     down     |  QSFP28E3-9  |
| Ethernet4_1  |    down     |     down     | QSFP28E4-13  |
| Ethernet5_1  |    down     |     down     | QSFP28E5-17  |
| Ethernet6_1  |    down     |     down     | QSFP28E6-21  |
| Ethernet7_1  |    down     |     down     | QSFP28E7-25  |
| Ethernet8_1  |    down     |     down     | QSFP28E8-29  |
| Ethernet9_1  |    down     |     down     | QSFP28E9-33  |
| Ethernet10_1 |    down     |     down     | QSFP28E10-37 |
| Ethernet11_1 |    down     |     down     | QSFP28E11-41 |
| Ethernet12_1 |    down     |     down     | QSFP28E12-45 |
| Ethernet13_1 |    down     |     down     | QSFP28E13-49 |
| Ethernet14_1 |    down     |     down     | QSFP28E14-53 |
| Ethernet15_1 |    down     |     down     | QSFP28E15-57 |
| Ethernet16_1 |    down     |     down     | QSFP28E16-61 |
| Ethernet17_1 |    down     |     down     |  PIU1-E1-65  |
| Ethernet18_1 |    down     |     down     |  PIU1-E2-69  |
| Ethernet19_1 |    down     |     down     |  PIU2-E1-73  |
| Ethernet20_1 |    down     |     down     |  PIU2-E2-77  |
| Ethernet21_1 |     up      |      up      |  PIU3-E1-81  |
| Ethernet22_1 |     up      |      up      |  PIU3-E2-85  |
| Ethernet23_1 |    down     |     down     |  PIU4-E1-89  |
| Ethernet24_1 |    down     |     down     |  PIU4-E2-93  |
| Ethernet25_1 |    down     |     down     |  PIU5-E1-97  |
| Ethernet26_1 |    down     |     down     | PIU5-E2-101  |
| Ethernet27_1 |    down     |     down     | PIU6-E1-105  |
| Ethernet28_1 |    down     |     down     | PIU6-E2-109  |
| Ethernet29_1 |    down     |     down     | PIU7-E1-113  |
| Ethernet30_1 |    down     |     down     | PIU7-E2-117  |
| Ethernet31_1 |    down     |     down     | PIU8-E1-121  |
| Ethernet32_1 |    down     |     down     | PIU8-E2-125  |
+--------------+-------------+--------------+--------------+
>
> show running-config
interface Ethernet1_1
  admin-status up
  ufd 1 uplink
  quit
!
interface Ethernet2_1
  admin-status up
  ufd 1 downlink
  quit
!
interface Ethernet21_1
  admin-status up
  quit
!
interface Ethernet22_1
  admin-status up
  quit
!
transponder piu3
  admin-status up
  quit
!
ufd 1
  quit
!
!
>
> interface Ethernet1_1
Execute 'exit' to exit
> exit
root@localhost:~# bcmcmd ps
ps
                 ena/        speed/ link auto    STP                  lrn  inter   max   cut   loop
           port  link  Lns   duplex scan neg?   state   pause  discrd ops   face frame  thru?  back
       ce0( 68)  up     4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce1( 72)  up     4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce2( 76)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce3( 80)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce4( 96)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce5(106)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce6(110)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce7(114)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce8(118)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce9(122)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce10(126)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce11(130)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce12( 84)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce13( 88)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce14( 92)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce15(102)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce16( 38)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce17( 34)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce18( 46)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce19( 42)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce20( 54)  up     4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce21( 50)  up     4  100G  FD   SW  No   Forward          None    D    CR4  9122    No
      ce22( 62)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce23( 58)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce24(  5)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce25(  1)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce26( 13)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce27(  9)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce28( 21)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce29( 17)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce30( 29)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce31( 25)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No

drivshell>
root@localhost:~#
root@localhost:~# gscli
>
> interface Ethernet1_1
interface(Ethernet1_1)>
interface(Ethernet1_1)> shutdown
interface(Ethernet1_1)> q
>
> show in brief
+--------------+-------------+--------------+--------------+
|     name     | oper-status | admin-status |    alias     |
+--------------+-------------+--------------+--------------+
| Ethernet1_1  |    down     |     down     |  QSFP28E1-1  |
| Ethernet2_1  |   dormant   |      up      |  QSFP28E2-5  |
| Ethernet3_1  |    down     |     down     |  QSFP28E3-9  |
| Ethernet4_1  |    down     |     down     | QSFP28E4-13  |
| Ethernet5_1  |    down     |     down     | QSFP28E5-17  |
| Ethernet6_1  |    down     |     down     | QSFP28E6-21  |
| Ethernet7_1  |    down     |     down     | QSFP28E7-25  |
| Ethernet8_1  |    down     |     down     | QSFP28E8-29  |
| Ethernet9_1  |    down     |     down     | QSFP28E9-33  |
| Ethernet10_1 |    down     |     down     | QSFP28E10-37 |
| Ethernet11_1 |    down     |     down     | QSFP28E11-41 |
| Ethernet12_1 |    down     |     down     | QSFP28E12-45 |
| Ethernet13_1 |    down     |     down     | QSFP28E13-49 |
| Ethernet14_1 |    down     |     down     | QSFP28E14-53 |
| Ethernet15_1 |    down     |     down     | QSFP28E15-57 |
| Ethernet16_1 |    down     |     down     | QSFP28E16-61 |
| Ethernet17_1 |    down     |     down     |  PIU1-E1-65  |
| Ethernet18_1 |    down     |     down     |  PIU1-E2-69  |
| Ethernet19_1 |    down     |     down     |  PIU2-E1-73  |
| Ethernet20_1 |    down     |     down     |  PIU2-E2-77  |
| Ethernet21_1 |     up      |      up      |  PIU3-E1-81  |
| Ethernet22_1 |     up      |      up      |  PIU3-E2-85  |
| Ethernet23_1 |    down     |     down     |  PIU4-E1-89  |
| Ethernet24_1 |    down     |     down     |  PIU4-E2-93  |
| Ethernet25_1 |    down     |     down     |  PIU5-E1-97  |
| Ethernet26_1 |    down     |     down     | PIU5-E2-101  |
| Ethernet27_1 |    down     |     down     | PIU6-E1-105  |
| Ethernet28_1 |    down     |     down     | PIU6-E2-109  |
| Ethernet29_1 |    down     |     down     | PIU7-E1-113  |
| Ethernet30_1 |    down     |     down     | PIU7-E2-117  |
| Ethernet31_1 |    down     |     down     | PIU8-E1-121  |
| Ethernet32_1 |    down     |     down     | PIU8-E2-125  |
+--------------+-------------+--------------+--------------+
>
> exit
root@localhost:~#
root@localhost:~# bcmcmd ps
ps
                 ena/        speed/ link auto    STP                  lrn  inter   max   cut   loop
           port  link  Lns   duplex scan neg?   state   pause  discrd ops   face frame  thru?  back
       ce0( 68)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce1( 72)  up     4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce2( 76)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce3( 80)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce4( 96)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce5(106)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce6(110)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce7(114)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce8(118)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce9(122)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce10(126)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce11(130)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce12( 84)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce13( 88)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce14( 92)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce15(102)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce16( 38)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce17( 34)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce18( 46)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce19( 42)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce20( 54)  up     4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce21( 50)  up     4  100G  FD   SW  No   Forward          None    D    CR4  9122    No
      ce22( 62)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce23( 58)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce24(  5)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce25(  1)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce26( 13)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce27(  9)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce28( 21)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce29( 17)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce30( 29)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
      ce31( 25)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No

drivshell>
root@localhost:~#
```

here we can see the ce1 and ce2 represent Ethernet1_1 and Ethernet2_1 respectively. When we disable Ethernet1_1 corresponding ce1 is in '!ena' state, but Ethernet2_1 (downlink) corresponding ce2 is still "up".

**Comment by ksandeep-palc** [comment]
@ishidawataru - Could you please share your inputs on the UFD functionality observed in this issue.

**Comment by ishidawataru** [comment]
> here we can see the ce1 and ce2 represent Ethernet1_1 and Ethernet2_1 respectively. When we disable Ethernet1_1 corresponding ce1 is in '!ena' state, but Ethernet2_1 (downlink) corresponding ce2 is still "up".

Looks like this doesn't match with the output of `bcmcmd ps`?

```
root@localhost:~# bcmcmd ps
ps
                 ena/        speed/ link auto    STP                  lrn  inter   max   cut   loop
           port  link  Lns   duplex scan neg?   state   pause  discrd ops   face frame  thru?  back
       ce0( 68)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce1( 72)  up     4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
       ce2( 76)  !ena   4  100G  FD   SW  No   Forward          None   FA    CR4  9122    No
```

`ce1` is `up` and `ce2` is `!ena`.

Also, it would be helpful if you can provide the log of `south-sonic`.

**Comment by prasadsriramps** [comment]
Hi @ishidawataru, 

apologies for confusing the ps output interface names, the corrected description would be, 

> here we can see the ce0 and ce1 represent Ethernet1_1 and Ethernet2_1 respectively. When we disable Ethernet1_1 corresponding ce0 is in '!ena' state, but Ethernet2_1 (downlink) corresponding ce1 is still "up".

attaching the south sonic logs, relevant to interface Ethernet1_1 shut and no shut. 

[south.log](https://github.com/oopt-goldstone/goldstone-mgmt/files/11832323/south.log)

**Comment by ishidawataru** [comment]
Thanks. I checked the code, and it looks like the logic to shut down downlinks when the uplink oper-status goes down is missing in `south-sonic`.

As you pointed out, the feature is incomplete right now.

**Comment by ishidawataru** [comment]
closed by #78
